### PR TITLE
Experiment: `DataList` - `DetailsList` rendering without measurements

### PR DESCRIPTION
**Pull request by cschleiden**
# No intention to merge in this form, want to start a discussion, will add some more details later

# Motivation

We've observed that the `DetailsList` causes a lot of layout thrashing and reflow even for the initial rendering. It happens because we measure the viewport, we measure every page after we have rendered it, and we render & measure the initial pages twice.

## Approach

1. No measurement for the initial render, use smart defaults or have the user pass in exact measurements
   - Either the user has to pass in the initial viewport height, **or** we default to `window.innerHeight`, which doesn't cause layout
   - For pages and items, we have the user pass in the height of a row. This makes it impossible/harder to support multi-line rows, but allows rendering without measuring anything
2. No measurements for columns, rely on flexbox to drop columns without enough horizontal space
3. No `pages` for virtualized list, reducing unwanted animations for example in grouped lists or tree controls built on `DetailsList`

# Benchmarking

For `DetailsList` I've used this pen:
https://codepen.io/cschleiden/pen/ayEXmE?editors=1000

For `DataList` I'ved a setup like this: 
https://codepen.io/cschleiden/pen/WEdLJB?editors=1000
(served the experiments bundle from localhost)

In both examples I render 100 items with 3 columns with the same display size. Reloaded each page four times and took the median:

| | `DataList` | `DetailsList` |
| ---| ---: |---: |
| Chrome - Median | **25.5ms** | 101ms |
| IE - Median | tbd | tbd |
| Edge - Median | tbd | tbd |

image

image

# Going forward

There are a number of options: 

### Pro
- Opt-in experience for consumers, don't need to be compatible with `DetailsList`
- Can keep the `DataList` without styling, making it easier for consumers to bring their own style (requirement for VSTS)

### Cons
- Duplicate work, we probably want to support single/multi selection, a11y etc. that `DetailsList` already provides

## Extend `DetailsList` with additional props 

Bring items 1, 2 and potentially 3 from Approach into `DetailsList`.

### Pro

- Potentially easier for consumers to opt-in, maybe just a prop?
- Everyone would be able to take advantage of improvements

### Cons
- `DetailsList` is already very heavy, adding even more props would bloat the interface
- Danger of breaking `DetailsList` consumers
- `List` is already complex, hard to make changes

### Pro
- Would give us some of the perf benefits, without rebuilding `DetailsList` completely 

### Cons
- Wouldn't give us benefits of not measuring columns (could maybe be added with another layout mode?)

### CODE REVIEW

File: packages/experiments/src/demo/index.tsx [hunk 1]
```diff
@@ -36,9 +37,10 @@ function _onLoad() {
 
   ReactDOM.render(
     <Fabric>
-      <Router onNewRouteLoaded={ _scrollAnchorLink }>
+      {/*<Router onNewRouteLoaded={ _scrollAnchorLink }>
```
Comment (dzearing) [inline packages/experiments/src/demo/index.tsx]: hmm this doesn't look expected

Comment (cschleiden) [inline packages/experiments/src/demo/index.tsx]: "-- No intention to merge in this format, want to start a discussion, will add some more details later --"

Comment (cschleiden) [inline packages/experiments/src/demo/index.tsx]: :)

File: packages/experiments/src/components/DataList/VirtualizedList.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    // TODO: CS: Can we use intersectionobserver here?
+    requestAnimationFrame(() => {
+      let scrollTop = 0;
+      if (this.scrollContainer as any === window) {
+        scrollTop = (this.scrollContainer as any).scrollY;
+      } else {
+        scrollTop = this.scrollContainer.scrollTop;
+      }
+
+      this.setState({
```
Comment (dzearing) [inline packages/experiments/src/components/DataList/VirtualizedList.tsx]: careful. Asynchronous operations can execute after a component has been unmounted, leading to a number of issues here like this.scrollContainer being null. Use `this._async.requestAnimationFrame` instead which will avoid executing the method after being unmounted.

File: packages/experiments/src/components/DataList/VirtualizedList.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    // TODO: Make element overflow auto, works better with how people use the list
+    return <div
+      className={ css(styles.root/*, hasExternalScrollContainer && styles.rootOverflow*/, className) }
+      ref={ this._resolveRef('root') }>
+      { this.renderItems() }
+    </div>;
+  }
+
+  private renderItems(): (JSX.Element | null)[] {
```
Comment (dzearing) [inline packages/experiments/src/components/DataList/VirtualizedList.tsx]: privates should be underscored (hmm this should be a lint rule)

File: packages/experiments/src/components/DataList/VirtualizedList.tsx (hunk 2 above)
Comment (cschleiden) [inline packages/experiments/src/components/DataList/VirtualizedList.tsx]: good point

**Pull request closed without merging**